libkbfs/disk_block_cache_wrapped.go

```diff
@@ -112,23 +111,48 @@ func (cache *diskBlockCacheWrapped) IsSyncCacheEnabled() bool {
 }
 
 // Get implements the DiskBlockCache interface for diskBlockCacheWrapped.
-func (cache *diskBlockCacheWrapped) Get(ctx context.Context, tlfID tlf.ID,
-	blockID kbfsblock.ID) (
+func (cache *diskBlockCacheWrapped) Get(
+	ctx context.Context, tlfID tlf.ID, blockID kbfsblock.ID,
+	preferredCacheType DiskBlockCacheType) (
 	buf []byte, serverHalf kbfscrypto.BlockCryptKeyServerHalf,
 	prefetchStatus PrefetchStatus, err error) {
 	cache.mtx.RLock()
 	defer cache.mtx.RUnlock()
 	primaryCache := cache.workingSetCache
 	secondaryCache := cache.syncCache
-	if cache.config.IsSyncedTlf(tlfID) && cache.syncCache != nil {
-		primaryCache, secondaryCache = secondaryCache, primaryCache
+	if preferredCacheType == DiskBlockSyncCache {
+		if cache.syncCache != nil {
+			primaryCache, secondaryCache = secondaryCache, primaryCache
+		} else {
+			log := cache.config.MakeLogger("DBC")
+			log.Warning("Sync cache is preferred, but there is no sync cache")
+		}
 	}
 	// Check both caches if the primary cache doesn't have the block.
 	buf, serverHalf, prefetchStatus, err =
-		primaryCache.Get(ctx, tlfID, blockID)
+		primaryCache.Get(ctx, tlfID, blockID, preferredCacheType)
 	if _, isNoSuchBlockError := err.(NoSuchBlockError); isNoSuchBlockError &&
 		secondaryCache != nil {
-		return secondaryCache.Get(ctx, tlfID, blockID)
+		buf, serverHalf, prefetchStatus, err = secondaryCache.Get(
+			ctx, tlfID, blockID, preferredCacheType)
+		if err != nil {
+			return nil, kbfscrypto.BlockCryptKeyServerHalf{}, NoPrefetch, err
+		}
+		if preferredCacheType != DiskBlockAnyCache {
```

Comment:
It seems to me that by using `DiskBlockAnyCache` everywhere, then if a block was once synced (and then later unsynced), we won't ever move it out of the sync cache. Is that desired behavior?

Reply:
If syncing on a TLF is completely unsynced, all the blocks of that TLF are removed from the sync cache.  I still need to work out a strategy for deleting blocks corresponding to unsynced partial paths -- there's a ticket for that and I'll handle it separately.

And if a block goes in the sync cache and the subsequently accessed with `DiskBlockAnyCache`, I think it's correct to leave it in the sync cache, since that was the most recent preferred cache for it.

But let me know if I misunderstood the question.

Reply:
I guess I'm saying that we never use `DiskBlockWorkingSetCache`, and it's unclear to me whether we ever plan on using it. And the fact that we use `DiskBlockAnyCache` in any situation where we even might use `DiskBlockWorkingSetCache`, means that we don't ever get into a situation where a block is moved from the sync cache to the working set cache (but we do get the reverse).

Reply:
Correct.  In fact I just pushed a fix that avoids deleting a synced block if `DiskBlockAnyCache` is passed in, because that was a bug.  I have no use case in mind for `DiskBlockWorkingSetCache` -- I can delete that constant if you find it confusing.

Reply:
No, it's fine. Just wanted to make sure I understood everything.